pkg/providers/pulumi/pulumi.go

```diff
@@ -0,0 +1,179 @@
+package pulumi
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/helmfile/vals/pkg/api"
+	"github.com/helmfile/vals/pkg/log"
+)
+
+const (
+	defaultPulumiAPIEndpointURL = "https://api.pulumi.com"
+)
+
+type provider struct {
+	log                  *log.Logger
+	backend              string
+	pulumiAPIEndpointURL string
+	pulumiAPIAccessToken string
+	organization         string
+	project              string
+	stack                string
+}
+
+type pulumiState struct {
+	Deployment pulumiDeployment `json:"deployment"`
+}
+
+func (r *pulumiState) findResourceByLogicalName(resourceType string, resourceLogicalName string) *pulumiResource {
+	for _, resource := range r.Deployment.Resources {
+		if resource.ResourceType == resourceType && strings.HasSuffix(resource.URN, resourceLogicalName) {
+			return &resource
+		}
+	}
+	return nil
+}
+
+type pulumiDeployment struct {
+	Resources []pulumiResource `json:"resources"`
+}
+
+type pulumiResource struct {
+	URN          string          `json:"urn"`
+	Custom       bool            `json:"custom"`
+	ID           string          `json:"id"`
+	ResourceType string          `json:"type"`
+	Inputs       json.RawMessage `json:"inputs"`
+	Outputs      json.RawMessage `json:"outputs"`
+}
+
+func (r *pulumiResource) getAttributeValue(resourceAttribute string, resourceAttributePath string) string {
+	var attributeValue gjson.Result
+	switch resourceAttribute {
+	case "inputs":
+		attributeValue = gjson.GetBytes(r.Inputs, resourceAttributePath)
+	case "outputs":
+		attributeValue = gjson.GetBytes(r.Outputs, resourceAttributePath)
+	}
+	return attributeValue.String()
+}
+
+func New(l *log.Logger, cfg api.StaticConfig, backend string) *provider {
+	p := &provider{
+		log:     l,
+		backend: backend,
+	}
+
+	if cfg.Exists("pulumi_api_endpoint_url") {
+		p.pulumiAPIEndpointURL = cfg.String("pulumi_api_endpoint_url")
+	} else {
+		p.pulumiAPIEndpointURL = os.Getenv("PULUMI_API_ENDPOINT_URL")
+		if p.pulumiAPIEndpointURL == "" {
+			p.pulumiAPIEndpointURL = defaultPulumiAPIEndpointURL
+		}
+	}
+
+	p.pulumiAPIAccessToken = os.Getenv("PULUMI_ACCESS_TOKEN")
+
+	if cfg.Exists("organization") {
+		p.organization = cfg.String("organization")
+	} else {
+		p.organization = os.Getenv("PULUMI_ORGANIZATION")
+	}
+
+	p.project = cfg.String("project")
+	p.stack = cfg.String("stack")
+
+	p.log.Debugf("pulumi: backend=%q, api_endpoint=%q, organization=%q, project=%q, stack=%q",
+		p.backend, p.pulumiAPIEndpointURL, p.organization, p.project, p.stack)
+
+	return p
+}
+
+func (p *provider) GetString(key string) (string, error) {
+	tokens := strings.Split(key, "/")
```

Comment:
@agershman we should check the length of tokens here. WDYT?

Reply:
Yep, I think that'd be helpful to be more explicit if the provided key is not in a supported format. Will address.